Add support for audio messages.

moldovaniosif opened the pull request:
<!--
Thanks for sending a pull request! Please make sure you click the link above to view the contribution guidelines, then fill out the blanks below.
-->

What does this implement/fix? Explain your changes.
---------------------------------------------------
Add support for audio messages.

Does this close any currently open issues?
------------------------------------------
No.


Any relevant logs, error output, etc?
-------------------------------------
<!--
If the logs is quite long, please paste to https://ghostbin.com/ and insert the link here.
-->

Any other comments?
-------------------
…

Where has this been tested?
---------------------------
**Devices/Simulators:** : iPhone 7, iPhone 4S, Simulator iPhone 7

**iOS Version:**  12.0, 9.3.5

**Swift Version:** 4.2

**MessageKit Version:** 1.0

moldovaniosif committed:
update to swift 4.2

moldovaniosif committed:
update MessageKit framework and tests to swift 4.2

moldovaniosif committed:
update podfile.lock

moldovaniosif committed:
Merge branch 'development' of https://github.com/MessageKit/MessageKit into audio_message_support

# Conflicts:
#	Cartfile.resolved
#	Example/ChatExample.xcodeproj/project.pbxproj
#	Example/Podfile.lock
#	Example/Sources/AppDelegate.swift
#	Example/Sources/ConversationViewController.swift
#	Example/Sources/Data Generation/SampleData.swift
#	Example/Sources/View Controllers/LaunchViewController.swift
#	MessageKit.xcodeproj/project.pbxproj
#	MessageKit.xcodeproj/xcshareddata/xcschemes/MessageKit.xcscheme
#	MessageKit.xcodeproj/xcshareddata/xcschemes/MessageKitTests.xcscheme
#	Sources/Controllers/MessagesViewController+Keyboard.swift
#	Sources/Views/InputBarItem.swift
#	Sources/Views/InputStackView.swift
#	Sources/Views/InputTextView.swift
#	Sources/Views/MessageInputBar.swift
#	Tests/ViewsTests/InputBarItemTests.swift

moldovaniosif committed:
partial development in add audio message support

moldovaniosif committed:
implement support for audio message

moldovaniosif committed:
fix configure audio cell

moldovaniosif committed:
Revert "fix configure audio cell"

This reverts commit da872d6d0ab73a1b10d8e6ee9d33780ae2253345, reversing
changes made to cda9cfa470c701a35f311f2b30c9a313c53538e5.

moldovaniosif committed:
fix compute audio cell size in the advance example view controller

moldovaniosif committed:
add test case for check return of audio message cell

nathantannar4 commented:
Excited to see someone submit a PR for this feature! 🔥 

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

moldovaniosif commented on Example/Sources/View Controllers/BasicExampleViewController.swift:
```diff
@@ -26,6 +26,7 @@ import UIKit
 import MapKit
 import MessageKit
 import MessageInputBar
+import AVFoundation
```
there's no need for this import

moldovaniosif commented on Example/Sources/View Controllers/ChatViewController.swift:
```diff
@@ -25,14 +25,15 @@ SOFTWARE.
 import UIKit
 import MessageKit
 import MessageInputBar
+import AVFoundation
```
there's no need for this import

moldovaniosif commented on Sources/Controllers/AudioController.swift:
```diff
@@ -0,0 +1,172 @@
+    /// The `MessageType` that is currently playing sound
+    fileprivate var playingMessage: MessageType?
+
+    fileprivate weak var messageCollectionView: MessagesCollectionView?
```
messageCollectionView property should not be optional.

moldovaniosif replied:
The messageCollectionView can be set a constant using let.

nathantannar4 commented on Documentation/FAQs.md:
```diff
@@ -1,6 +1,13 @@
 # MessageKit Frequently Asked Questions
 
-**Why doesn't the `MessageInputBar` appear in my controller?**
```
Thanks for updating the FAQ 💯

moldovaniosif replied:
I souldn't be credit for updating FAQ, I had no ideea why FAQ had been updated in my PR.

nathantannar4 commented on Example/ChatExample.xcodeproj/project.pbxproj:
```diff
@@ -676,7 +773,7 @@
 				ALWAYS_EMBED_SWIFT_STANDARD_LIBRARIES = YES;
 				ASSETCATALOG_COMPILER_APPICON_NAME = AppIcon;
-				DEVELOPMENT_TEAM = "";
+				DEVELOPMENT_TEAM = 5HU545WEHV;
```
Please remove development team signing

nathantannar4 commented on Example/Sources/Data Generation/SampleData.swift:
```diff
@@ -174,15 +183,6 @@ final internal class SampleData {
-            messages.append(message)
-        }
-        completion(messages)
-    }
```
Is there a reason you removed this?

moldovaniosif replied:
This method is not used any more, there is other method " func getMessages(count: Int, completion: ([MockMessage]) -> Void) " that does the same thing.

nathantannar4 commented on Example/Sources/Models/MockMessage.swift:
```diff
@@ -53,6 +54,22 @@ private struct ImageMediaItem: MediaItem {
+        self.url = url
+        self.size = CGSize(width: 160, height: 35)
+        // compute duration
+        let audioAsset = AVURLAsset.init(url: url)
```
`let audioAsset = AVURLAsset(url: url)`

nathantannar4 commented on Sources/Controllers/AudioController.swift:
```diff
@@ -0,0 +1,172 @@
+            case .audio(let item):
+                playingCell = audioCell
+                playingMessage = message
+                audioPlayer = try? AVAudioPlayer.init(contentsOf: item.url)
```
`audioPlayer = try? AVAudioPlayer(contentsOf: item.url)`

nathantannar4 commented on Sources/Controllers/AudioController.swift:
```diff
@@ -0,0 +1,172 @@
+    }
+
+    internal func stopAudioInCurrentPlayingCell() {
+        // call stop audio delegate
```
If the audio player is nil then we don't need to go through the stopping logic
```
// call stop audio delegate
guard let player = audioPlayer else { return }
if let cell = playingCell {
             cell.delegate?.didStopAudio(in: cell)
         }
audioPlayer.stop()
audioPlayerDidFinishPlaying(audioPlayer, successfully: true)
```
Also remove comments, the code is self explanatory

nathantannar4 commented on Sources/Controllers/AudioController.swift:
```diff
@@ -0,0 +1,172 @@
+// MARK: - AVAudioPlayerDelegate
+extension AudioController: AVAudioPlayerDelegate {
+
+    internal func audioPlayerDidFinishPlaying(_ player: AVAudioPlayer, successfully flag: Bool) {
```
`flag: Bool` is not used, also why are we passing in a `player: AVAudioPlayer` parameter when we have access to the `audioPlayer` in the `AudioController`

moldovaniosif replied:
The audioPlayerDidFinishPlaying(::) is a delegate method that notify when AVAudioPlayer finished play. I can not modify the structure of this method since this one is from the OS system. I think is better to implement a new method without parameters an call it from audioPlayerDidFinishPlaying(::) and stopAudioInCurrentPlayingCell() methods.

nathantannar4 commented on Sources/Controllers/MessagesViewController.swift:
```diff
@@ -71,12 +71,19 @@ UICollectionViewDelegateFlowLayout, UICollectionViewDataSource {
 
     internal var selectedIndexPathForMenu: IndexPath?
 
+    internal var audioController: AudioController!
```
I think this should be open to allow people to set their own `AudioController`

moldovaniosif replied:
You are right, I wasn't thinking about how other users can customize the AudioController.

nathantannar4 commented on Sources/Controllers/AudioController.swift:
```diff
@@ -0,0 +1,172 @@
+import UIKit
+import AVFoundation
+
+internal class AudioController: NSObject {
```
Make `AudioController` a protocol to allow people to set their own `AudioController` in the `MessagesViewController`, then make this an open class like `BasicAudioController`

This would allow people to customize  the experience to which is one of the MessageKit goals. We try to make everything open/public unless it will break the framework

moldovaniosif replied:
@nathantannar4 greate ideea.

nathantannar4 commented on Sources/Protocols/AudioCellDelegate.swift:
```diff
@@ -0,0 +1,42 @@
+import Foundation
+
+/// A protocol used by `AudioMessageCell` to notify about taps on the play/pause button.
+internal protocol AudioCellDelegate: AnyObject {
```
`public`, see earlier comments

moldovaniosif replied:
I've remove the `AudioCellDelegate` protocol and move the didTapPlayButton() delegate method in the `MessageCellDelegate` protocol. In this way the user can take any action when audio play button was pressed.

nathantannar4 commented on Sources/Views/Cells/AudioMessageCell.swift:
```diff
@@ -0,0 +1,171 @@
+        let pauseImage = AudioMessageCell.getImageWithName("pause")
+        playButton.setImage(playImage?.withRenderingMode(.alwaysTemplate), for: .normal)
+        playButton.setImage(pauseImage?.withRenderingMode(.alwaysTemplate), for: .selected)
+
```
remove newline

nathantannar4 commented on Sources/Views/Cells/AudioMessageCell.swift:
```diff
@@ -0,0 +1,171 @@
+    /// The play button view to display on audio messages.
+    open lazy var playButton: UIButton = {
+        let playButton = UIButton.init(type: .custom)
+        let playImage = AudioMessageCell.getImageWithName("play")
```
Should use an enum here rather than a raw string

nathantannar4 commented on Sources/Views/Cells/AudioMessageCell.swift:
```diff
@@ -0,0 +1,171 @@
+    open lazy var progressView: UIProgressView = {
+        let progressView = UIProgressView.init(progressViewStyle: .default)
+        progressView.progress = 0.0
+
```
remove newline

nathantannar4 commented on Sources/Views/Cells/AudioMessageCell.swift:
```diff
@@ -0,0 +1,171 @@
+        durationLabel.textAlignment = .right
+        durationLabel.font = UIFont.systemFont(ofSize: 14)
+        durationLabel.text = "0:00"
+
```
remove newline

nathantannar4 commented on Sources/Views/Cells/AudioMessageCell.swift:
```diff
@@ -0,0 +1,171 @@
+    }()
+
+    // MARK: - Internal Methods
+    internal class func durationString(from duration: Float) -> String {
```
Not sure a `class` func is the right way to approach this, is there a better way? At the very least this should be `open`

moldovaniosif replied:
@nathantannar4 you are right, this class method is not the best approach. I had removed it from AudioMessageCell and instead I created a new delegate methods on MessageDisplayDelegate. In this case user can proide it's own implementation for formatting the audio sound duration or can use the default implementation.

nathantannar4 reviewed (changes requested):
A lot of changes to look though here. General feedback is that I would like more classes/methods to be `open` so they are overridable/subclassable. That way users can customize it. The design of the `AudioMessageCell` is great! And seems to work without issue.

moldovaniosif committed:
apply review changes

moldovaniosif commented:
@nathantannar4 thank you for yhe details review. I had apply the changes.

nathantannar4 commented on Example/Sources/View Controllers/ChatViewController.swift:
```diff
@@ -207,6 +207,20 @@ extension ChatViewController: MessageCellDelegate {
+                return
+        }
+        // check if should play sound or pause
+        if cell.playButton.isSelected == true { // sound is playing - prepare to pause sound
```
If this is how playing sound is determined its not a very intuitive way. I think something like `isPlaying: Bool` should be accessible in the AudioController as its what controls the sound, not the cell

nathantannar4 commented on Sources/Controllers/BasicAudioController.swift:
```diff
@@ -25,55 +25,79 @@
 
     /// The `AudioMessageCell` that is currently playing sound
-    fileprivate var playingCell: AudioMessageCell?
+    open var playingCell: AudioMessageCell?
```
I think this should be weak

moldovaniosif replied:
You are right.

nathantannar4 commented on Sources/Controllers/BasicAudioController.swift:
```diff
@@ -25,55 +25,79 @@
 
     /// The `MessageType` that is currently playing sound
-    fileprivate var playingMessage: MessageType?
+    open var playingMessage: MessageType?
```
I think this should be weak

moldovaniosif replied:
The `playingMessage ` can not be made weak because `MessageType` is not a class bound object.

nathantannar4 commented on Sources/Controllers/BasicAudioController.swift:
```diff
@@ -25,55 +25,79 @@
 
-    fileprivate weak var messageCollectionView: MessagesCollectionView?
+    // The `MessagesCollectionView` where the playing cell exist
+    public let messageCollectionView: MessagesCollectionView
```
This should defiantly be weak

`open weak var messageCollectionView: MessagesCollectionView!`

nathantannar4 commented on Sources/Controllers/BasicAudioController.swift:
```diff
@@ -25,55 +25,79 @@
         if playingMessage?.messageId == message.messageId { // resume curently pause audio
             self.resumeSound()
         } else {
             if playingCell != nil { // there is a sound on other cell that is already playing or in pause - prepare to stop playing sound and play the sound for receive message
```
I don't think this is a good way of determining if a sound is playing. As earlier mentioned the AudioController should have an accessor for if its currently playing

nathantannar4 commented on Sources/Controllers/BasicAudioController.swift:
```diff
@@ -25,55 +25,79 @@
                 audioPlayer?.prepareToPlay()
+                audioPlayer?.delegate = self
                 audioPlayer?.play()
                 audioCell.playButton.isSelected = true  // show pause button on audio cell since the sound is currently playing
```
As before mentioned, not a good way of managing `isPlaying`

nathantannar4 commented on Sources/Controllers/BasicAudioController.swift:
```diff
@@ -95,31 +124,46 @@ internal class AudioController: NSObject {
+                    fatalError(MessageKitError.nilMessagesDisplayDelegate)
+                }
+                cell.durationLabel.text = displayDelegate.formatDuration(Float(player.duration), for: cell, in: messageCollectionView)
             }
```
I think that if the currentMessage has changed and is no longer the stored message audio playing and the timer should terminate

nathantannar4 commented on Sources/Controllers/BasicAudioController.swift:
```diff
@@ -143,30 +187,23 @@ internal class AudioController: NSObject {
         }
     }
 
+    private func createAudioPlayer(url: URL) -> AVAudioPlayer? {
```
This can be removed and where needed use:
```
guard let audioPlayer = audioPlayer = try? AVAudioPlayer(contentsOf: url) else { return }
```
Since if we can't create the player we just don't play anything rather than returning nil and going about as if we did create a player

nathantannar4 commented on Sources/Controllers/BasicAudioController.swift:
```diff
@@ -95,31 +124,46 @@ internal class AudioController: NSObject {
-    internal func stopAudioInCurrentPlayingCell() {
-        // call stop audio delegate
+    /// Stops any ongoing audio playing if exists
+    open func stopAnyOngoingPLaying() {
```
typo

nathantannar4 commented on Sources/Controllers/BasicAudioController.swift:
```diff
@@ -143,30 +187,23 @@ internal class AudioController: NSObject {
+
+    open func audioPlayerDidFinishPlaying(_ player: AVAudioPlayer, successfully flag: Bool) {
+        self.stopAnyOngoingPLaying()
     }
```
I would add the other delegate methods for the case where there is some audio playing corruption

nathantannar4 commented on Sources/Controllers/BasicAudioController.swift:
```diff
@@ -143,30 +187,23 @@ internal class AudioController: NSObject {
-        audioPlayer = nil
-        playingMessage = nil
-        playingCell = nil
+extension BasicAudioController: AVAudioPlayerDelegate {
```
Dont make this an extension, it means the methods cannot be overridden even though they are open

nathantannar4 commented on Sources/Controllers/MessagesViewController.swift:
```diff
@@ -60,6 +60,9 @@ UICollectionViewDelegateFlowLayout, UICollectionViewDataSource {
     }
 
+    /// The `BasicAudioController` controll the AVAudioPlayer state (play, pause, stop) and udpate audio cell UI accordingly.
+    open var audioController: BasicAudioController!
```
Make this lazy, as those who don't use audio messages will never need it to be created
`open lazy var audioController = BasicAudioController(messageCollectionView: messageCollectionView)`

nathantannar4 commented on Sources/Protocols/MessagesDisplayDelegate.swift:
```diff
@@ -184,6 +173,33 @@ public protocol MessagesDisplayDelegate: AnyObject {
+    ///     1. return the time as 0:ss if duration is up to 59 seconds                         (e.g. 0:03     means 0 minutes and 3 seconds)
+    ///     2. return the time as m:ss if duration is greater than 59 and lower than 3600      (e.g. 12:23    means 12 mintues and 23 seconds)
+    ///     3. return the time as h:mm:ss for anything longer that 3600 seconds                (e.g. 1:19:08  means 1 hour 19 minutes and 8 seconds)
+    func formatDuration(_ duration: Float, for audioCell: AudioMessageCell, in messageCollectionView: MessagesCollectionView) -> String
```
`audioProgressTextFormat` is a bit more discoverable/descriptive

nathantannar4 commented on Sources/Views/Cells/AudioMessageCell.swift:
```diff
@@ -160,12 +138,16 @@ open class AudioMessageCell: MessageContentCell {
     // MARK: - Helpers
 
-    private class func getImageWithName(_ imageName: String) -> UIImage? {
+    private class func getImageWithName(_ imageName: ImageName) -> UIImage? {
```
This should be open so people could subclass the cell and get the image we have

nathantannar4 commented on Sources/Views/Cells/AudioMessageCell.swift:
```diff
@@ -160,12 +138,16 @@ open class AudioMessageCell: MessageContentCell {
         return image
     }
 
+    private enum ImageName: String {
```
Public and place at the top of the class

nathantannar4 reviewed (changes requested):
Thanks for making the previous changes! I think some improvements could still be made

moldovaniosif committed:
apply second review changes

moldovaniosif commented:
Second review changes had been applied.

moldovaniosif commented:
Hi @nathantannar4 is everything ok with this latest changes?
Is there anything that could be improved?

nathantannar4 commented on Sources/Controllers/BasicAudioController.swift:
```diff
@@ -25,21 +25,37 @@
+    case pause
+
+    /// The audio controller is not playing any sound and audioPlayer is nil
+    case none
```
Perhaps `stopped` is a better name for this, a `none` state doesn't make sense grammatically

nathantannar4 commented on Sources/Views/Cells/AudioMessageCell.swift:
```diff
@@ -0,0 +1,152 @@
+    }()
+
+    open lazy var progressView: UIProgressView = {
+        let progressView = UIProgressView.init(progressViewStyle: .default)
```
Please cleanup these `.init` as its not needed

nathantannar4 commented on Sources/Views/Cells/AudioMessageCell.swift:
```diff
@@ -0,0 +1,152 @@
+        super.prepareForReuse()
+        progressView.progress = 0
+        playButton.isSelected = false
+        self.durationLabel.text = "0:00"
```
Please remove `self` whenever possible

nathantannar4 commented on Sources/Controllers/BasicAudioController.swift:
```diff
@@ -0,0 +1,223 @@
+            // call start audio delegate
+            audioCell.delegate?.didStartAudio(in: audioCell)
+        default:
+            print("Error - can not play sound for a message that is not of kind .audio")
```
I believe we have a file for MessageKit errors, can you please move this there and follow the convention we use for printing errors?

nathantannar4 commented on Sources/Controllers/BasicAudioController.swift:
```diff
@@ -0,0 +1,223 @@
+
+    /// Resume a currently pause audio sound
+    open func resumeSound() {
+        guard let player = audioPlayer else {
```
We should also be guarding the `cell` here

nathantannar4 reviewed (changes requested):
Getting to a much better state! In general, can you please cleanup your commenting so things are more readable. Please note that as this is a larger change we won't be able to include it in v1.1. It would be great if others could also review this, I still haven't had time to test how customizable this is. 

It would also be great if there were some tests for the AudioController 😊 

Thanks for your continued contribution

moldovaniosif committed:
Merge branch 'development' of https://github.com/MessageKit/MessageKit into tagonsoft_development

# Conflicts:
#	Example/Podfile.lock
#	Example/Sources/View Controllers/ChatViewController.swift
#	MessageKit.xcodeproj/project.pbxproj
#	Sources/Protocols/MessageCellDelegate.swift

moldovaniosif committed:
clean code and add tests for audio controller

moldovaniosif commented:
I've added some test cases for AudioController. If i've miss some cases please let me know.
I had no ideea why circleUI tests are falling, on my mashine are allways pass. It may be because I've added a resource to test target?

nathantannar4 commented:
CircleCI probably complaining about xcode/Swift version. Did you migrate this PR to Swift 4.2 yet? You might need to rebase it on the latest `development` commit 

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

SD10 commented on Sources/Protocols/AudioControllerDelegate.swift:
```diff
@@ -0,0 +1,45 @@
+    ///
+    /// - Note:
+    ///   This protocol method is called by MessageKit every time an audio cell needs to be configure
+    func configureCell(_ cell: AudioMessageCell, message: MessageType)
```
Do you mind naming this `configureAudioCell`? So that it's clearer as to what kind of cell

SD10 commented:
@moldovaniosif This is looking pretty cool ❤️ Thanks for all the effort 🥇 And @nathantannar4 thanks for not letting this slip through the cracks and doing the review. I reviewed it briefly now but plan to take a deeper look

moldovaniosif committed:
rename AudioControllerDelegate method

moldovaniosif committed:
add logs in AudioControllerTests

moldovaniosif commented:
Current PR is using swift 4.2 and yesterday I rebased it to the lastest development commit. CircleCI is complaing to the test cases I've just added in the AudioControllerTests, some true asserts are falling. I don't know how CircleCI is working or how can debug it, if I runt tests from Xcode everything is working fine.

talanov commented on Tests/ControllersTest/AudioControllerTests.swift:
```diff
@@ -0,0 +1,155 @@
+        XCTAssertNotNil(audioController.audioPlayer)
+        XCTAssertNotNil(audioController.playingCell)
+        XCTAssertNotNil(audioController.playingMessage)
+        XCTAssertEqual(audioCell.durationLabel.text, "0:00")
```
```suggestion
```

moldovaniosif replied:
But why when I run the test from Xcode isPlaying returns true, and from CirecleCi interface returns false? I don't understant what is wrong or what is the diference between Xocde and CircleCI

talanov reviewed (changes requested):
Tests are failing.

audioPlayer isn't nil, yet, isPlaying returns false.

moldovaniosif committed:
apply talanov edit sugestion

talanov commented:
Yep, all tests are passed @ Xcode.

nathantannar4 commented:
@moldovaniosif I'd like to get this in the 2.1 release 😊 There is some merge conflicts and we still need a CHANGELOG entry for this. You can place in under upcoming release as a breaking change

moldovaniosif committed:
Merge branch 'development' of https://github.com/MessageKit/MessageKit into tagonsoft_development

# Conflicts:
#	Example/Sources/Data Generation/SampleData.swift

moldovaniosif committed:
add Audio Message cell in Settings screen

moldovaniosif committed:
try fix Circle CI fail test

moldovaniosif committed:
try fix Circle CI tests

moldovaniosif committed:
try fix Circle CI tests

moldovaniosif committed:
update change log

moldovaniosif commented:
I had resolved the merge conflicts, update the change log and remove the aserts that breake the Circle CI.

SD10 commented on Sources/Protocols/AudioControllerDelegate.swift:
```diff
@@ -0,0 +1,45 @@
+/// A standard protocol representing an audio controller.
+/// Use this protocol to create your own audio controller object to be used by MessageKit.
+/// MessageKit provide a default implmentation for AudioControllerDelegate in BasicAudioController
+public protocol AudioControllerDelegate: AnyObject {
```
Does this need to be a separate delegate or can it be part of `MessagesDisplayDelegate`?

moldovaniosif replied:
The `AudioControllerDelegate` contains the mandatory methods which an AudioController must implement, so any one hwo want a custom implmentation should conform with this protocol. That's way i make it a separate protocol.
I was thinking because `BasicAudioController` updates the `AudioMessageCell` UI it should be part of MessageKit. But it can be moved outside the framework also. If you have any ideea of how the arhitecture for audio can be improved please let me know.

SD10 reviewed (commented):
@moldovaniosif Thanks for your hard work on this. I don't want it to get swept under the rug. Taking a bit of time to get around to it. Originally, I didn't want any audio player to be part of the library (just the cell UI). Can you make a case for why it needs to be? Thanks again 👍

moldovaniosif committed:
Merge branch 'development' of https://github.com/MessageKit/MessageKit into tagonsoft_development

# Conflicts:
#	CHANGELOG.md

moldovaniosif committed:
Merge branch 'development' of https://github.com/MessageKit/MessageKit into tagonsoft_development

# Conflicts:
#	CHANGELOG.md

moldovaniosif commented:
Hi, is there any thing else to change for this pull request? Should I move the AudioController outside the MessageKit?

nathantannar4 commented:
@moldovaniosif Im ok with having a basic Audio controller, but I think there still needs to be some more work on making a more protocol oriented design for it if we keep it, as your design only works for local audio clips, not an audio stream from a server URL. The UI is defiantly on point, perhaps the best way to go about this is if you take out the AudioController parts of this PR and just leave the UI? Then we can make a separate PR for the AudioController

iceman201 commented on Example/Sources/View Controllers/ChatViewController.swift:
```diff
@@ -64,6 +64,8 @@ class ChatViewController: MessagesViewController, MessagesDataSource {
     override func viewDidDisappear(_ animated: Bool) {
         super.viewDidDisappear(animated)
         MockSocket.shared.disconnect()
+
```
remove empty line

moldovaniosif committed:
Merge branch 'development' of https://github.com/MessageKit/MessageKit into tagonsoft_development

# Conflicts:
#	CHANGELOG.md

SD10 commented on Sources/Controllers/MessagesViewController.swift:
```diff
@@ -208,6 +211,11 @@ UICollectionViewDelegateFlowLayout, UICollectionViewDataSource {
+        case .audio:
+            let cell = messagesCollectionView.dequeueReusableCell(AudioMessageCell.self, for: indexPath)
+            cell.configure(with: message, at: indexPath, and: messagesCollectionView)
+            audioController.configureAudioCell(cell, message: message) // this is needed especily when the cell is reconfigure while is playing sound
```
@moldovaniosif Stopping by again because I gave this more thought tonight. The only thing that concerns me here is `BasicAudioController` from a maintenance perspective. The reason this feature has had a lower priority to me is because I will never use it personally.

1. Is there any way to add Audio cells to MessageKit without the burden of the player?
2. If there's no way around this, can you be the code owner for this feature of MessageKit?

moldovaniosif replied:
1. I think it should be possible to move the `AudioController` outside the framework.
2. Yes, I can take the responsibility for this feature.

When I will find some time I will try to pull out the `AudioController` as @nathantannar4 mention letting inside the framework only the audio cell UI. Then I can make a new pull request for the `AudioController`. If you have some thoughts about how the `AudioController` should behave or how should be structured I am opened to any suggestions.

Commit:
Merge branch 'development' of github.com:MessageKit/MessageKit into development

# Conflicts:
#	CHANGELOG.md

Commit:
pull out BasicAudioController from MessageKit framework

moldovaniosif commented:
I've moved the `BasicAudioController` from MessageKit framework in to the example app. Currently the framework only implement the audio cell UI while the app example treats playing the audio sound.

alexBotteri commented:
@zhongwuzw do you have an estimate date for this PR review? TY!

zhongwuzw commented:
@alexBotteri Emm, sorry, I may don't have much time to review this. cc @SD10 @nathantannar4 Hi guys, any thoughts about this `PR`?

nathantannar4 commented:
@zhongwuzw I've reviewed this quite heavily. Now that it's just the UI elements which are being added it's good. 

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

bmt2018 commented:
@SD10 @nathantannar4 @zhongwuzw any idea how far away an official release with this feature merged in might be?  I just started using MK and was trying to understand how to create a custom cell to support audio messages when I found this PR.  This is exactly what I need to implement but I'm not sure how I can leverage it until it's been merged into a release?

nathantannar4 commented:
@bmt2018 you can specify @moldovaniosif branch to use this

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

moldovaniosif committed:
apply code review and update cardfile.resolve
